### PR TITLE
Tezos: support the new protocol update (005-BABYLON)

### PR DESCRIPTION
**Pull request by adonagy**
Hi!

This PR updates the firmware to be compatible with the new tezos protocol (https://tezos.gitlab.io/mainnet/protocols/005_babylon.html). The firmware is kept backward compatible with the current protocol to ensure proper functionality in case the protocol switch does not occur (not enough votes).

With these changes to the protocol, the binary representation of the operations have changed.

A major change is the deprecation of the scriptless kt1 accounts mainly used for token delegation. Now users can delegate tokens from the implicit (tz1, tz2 and tz3) accounts. After the protocol is approved and switched to, all the scriptless kt accounts will contain a master.tz script. This script will allow manipulation with the before scriptless accounts (set the delegate or transfer the tokens to an implicit account), thus creating new kinds of transaction operations. These transactions have an entrypoint with MICHELSON code prepended. 

Tested on the tezos-zeronet running the new protocol. Tests are included in the corresponding test file.

### CODE REVIEW

**Comment by prusnak** [comment]
When is the launch of this hardfork expected?

**Comment by adonagy** [comment]
In approximately 17 days, if voting participation reaches the quorum and the proposal has a super-majority in favor. 

https://tzscan.io/proposals

**Comment by prusnak** [comment]
That's too bad. The first release we can fit this in is 6th of November ...

**Comment by adonagy** [comment]
Unfortunately, they are not. In the proposed update the tags of the operations (transaction, reveal, origination, delegations) are changed to, quote: "make the incompatibility explicit between the two protocols"

**Comment by matejcik** [comment]
is there a reason to support both protocols then?

**Comment by prusnak** [comment]
> Is there a reason to support both protocols then?

Good point. I think it makes sense to rework the PR so only the post-hardfork functionality is available.

**Comment by adonagy** [comment]
@matejcik @prusnak  Good point indeed.

The reason behind supporting both protocols was the synchronization between releases. In the case, that now is not possible, the firmware release happens before the protocol update

I'll rework the PR to support only the new protoco.

File: common/protob/messages-tezos.proto [hunk 1]
```diff
@@ -60,7 +61,7 @@ message TezosSignTx {
     /*
      * Tezos contract ID
      */
-    message TezosContractID {
+     message TezosContractID {
```
Comment (prusnak) [inline common/protob/messages-tezos.proto]: We can remove `TezosContractID` completely if it's not used anymore at all.

File: common/protob/messages-tezos.proto [hunk 2]
```diff
@@ -75,7 +76,7 @@ message TezosSignTx {
      * Structure representing information for reveal
      */
     message TezosRevealOp {
-        optional TezosContractID source = 1;
+        optional bytes source = 1;
```
Comment (prusnak) [inline common/protob/messages-tezos.proto]: Please change also the field number if you change the type.

`optional bytes source = 7;`

Same for other changes below.

File: common/protob/messages-tezos.proto (hunk 1 above)
Comment (adonagy) [reply common/protob/messages-tezos.proto]: We still use `TezosContractID` in the `TezosTransactionOp` as the type of the destination field

File: common/protob/messages-tezos.proto (hunk 2 above)
Comment (adonagy) [reply common/protob/messages-tezos.proto]: Will do.

File: common/protob/messages-tezos.proto (hunk 1 above)
Comment (prusnak) [reply common/protob/messages-tezos.proto]: What's the difference between Implicit and Originated contract type?

Comment (adonagy) [reply common/protob/messages-tezos.proto]: Originated accounts(KT1) are created by the origination operation from an implicit account. The originated account can contain MICHELSON code, they're essentially smart contracts. 

Implicit accounts(tz{1, 2, 3}) are the 'classic' accounts, that are holding tokens and they can not include code (contract).

The contract type here is referring to the account type.

**Comment by prusnak** [comment]
OK, so it seems the network has switched to Babylon.

Can you please test whether the firmware works with the new hardfork?

File: common/protob/messages-tezos.proto [hunk 3]
```diff
@@ -60,7 +60,7 @@ message TezosSignTx {
     /*
      * Tezos contract ID
      */
-    message TezosContractID {
+     message TezosContractID {
```
Comment (matejcik) [inline common/protob/messages-tezos.proto]: this indent should be fixed

File: core/src/trezor/messages/TezosTransactionOp.py [hunk 4]
```diff
@@ -3,6 +3,12 @@
 import protobuf as p
 
 from .TezosContractID import TezosContractID
+from .TezosTransactionSmartContractDelegationOp import (
```
Comment (matejcik) [inline core/src/trezor/messages/TezosTransactionOp.py]: not sure what happened here but this shouldn't be produced by `make gen`

**Comment by adonagy** [comment]
> Can you please test whether the firmware works with the new hardfork?
 I'll test the firmware on the the updated Babylon protocol on the mainnet.  

In the meantime I've improved the the added operations, made it more readable and better to understand. I will update the PR ASAP, most likely tomorrow. I'll include the updated tests as well.

**Comment by prusnak** [comment]
Great! I want to merge this code for our QA to test, so it would be best if we had the final changes on Wednesday the latest.

File: core/src/apps/tezos/sign_tx.py [hunk 5]
```diff
@@ -215,3 +253,77 @@ def _encode_ballot(w: bytearray, ballot):
...
+    MICHELSON_LENGTH = 42  # length is fixed this time(no variable length fields)
+
+    _encode_kt_common(w, MICHELSON_LENGTH, "PUSH")
+    write_bytes(w, smart_contract_delegation.delegate)
+    write_bytes(w, bytes(helpers.MICHELSON_INSTRUCTION_BYTES["SOME"]))
+    write_bytes(w, bytes(helpers.MICHELSON_INSTRUCTION_BYTES["SET_DELEGATE"]))
+    write_bytes(w, bytes(helpers.MICHELSON_INSTRUCTION_BYTES["CONS"]))
+
+
+def _encode_smart_contract_delegation_remove(w: bytearray, smart_contract_delegation):
```
Comment (matejcik) [inline core/src/apps/tezos/sign_tx.py]: this function does not use the `smart_contract_delegation` argument. it should be removed

File: core/src/apps/tezos/sign_tx.py [hunk 6]
```diff
@@ -215,3 +253,77 @@ def _encode_ballot(w: bytearray, ballot):
...
+        write_uint8(w, 128 | byte)
+        _encode_zarith(w, modified)
+
+
+def _encode_kt_common(w: bytearray, sequence_length, operation):
+    ADDRESS_LENGTH = 21
+
+    argument_length = (
+        sequence_length + 5
+    )  # 5 = tag and sequence_length (1 byte + 4 bytes)
```
Comment (matejcik) [inline core/src/apps/tezos/sign_tx.py]: please move the comment on the preceding line so that this doesn't happen

**Comment by adonagy** [comment]
I've updated the PR with the refactor, added the last operation and tests. The checks are failing because the generated files seem little off. I'm allways using `make gen` for file regeneration. Don't know what's causing the problem. 

> I didn't understand the point of _encode_natural function, but I assume protocol-wise you know that it works ;)

It is almost identical with the zarith encoding. The difference being in setting the 7th bit if the value is negative. In the transfers however we never send negative amount of tokens. So the bit is never set but is still reserved for the sign.

File: Pipfile [hunk 7]
```diff
@@ -64,4 +64,4 @@ attrs = "==18.2.0"  # let's use this version because of the nixos pipenv/pip iss
 scan-build = "*"
 
 [pipenv]
-allow_prereleases = true
+allow_prereleases = true
```
Comment (matejcik) [inline Pipfile]: you're not leaving newlines at end of files. fix your editor config ;)

File: core/src/apps/tezos/sign_tx.py [hunk 8]
```diff
@@ -21,14 +21,43 @@
...
+            parameters_manager = msg.transaction.parameters_manager
+
+            # operation to delegate from a smart contract with manager.tz
+            if parameters_manager.set_delegate is not None:
+                delegate = _get_address_by_tag(parameters_manager.set_delegate)
+                await layout.require_confirm_delegation_baker(ctx, delegate)
+                await layout.require_confirm_set_delegate(ctx, msg.transaction.fee)
+
+            # operation to remove delegate from the smart contract with manager.tz
+            if parameters_manager.cancel_delegate is not None:
```
Comment (matejcik) [inline core/src/apps/tezos/sign_tx.py]: does this mean that you can set and cancel delegate and perform a transfer in a single transaction?

**Comment by matejcik** [comment]
> I've updated the PR with the refactor, added the last operation and tests. The checks are failing because the generated files seem little off. I'm allways using `make gen` for file regeneration. Don't know what's causing the problem.

ISTM you fixed conflicts manually after the rebase instead of simply running `make gen` again?
I can push a commit that fixes it, but presumably if you run `make gen` on your side with the same result.

**Comment by adonagy** [comment]
> ISTM you fixed conflicts manually after the rebase instead of running make gen again?
I can push a commit that fixes it, but presumably you can run make gen on your side with the same result.

Hmm, I'm quite sure I did run make gen after rebase, I'll double check. 

On an another note, whenever I run make gen the BITCOIN ONLY checks disappear in the core/src/trezor/messages/Capability.py and core/src/trezor/messages/Capability.py files and I have to fix the files manually.

File: Pipfile (hunk 7 above)
Comment (adonagy) [reply Pipfile]: Oh, right. Will do.

**Comment by matejcik** [comment]
> On an another note, whenever I run make gen the BITCOIN ONLY checks disappear in the core/src/trezor/messages/Capability.py and core/src/trezor/messages/Capability.py files and I have to fix the files manually.

The `BITCOIN_ONLY` check is added by a bunch of seds in `tools/build_protobuf`. `make gen` runs that. No idea why it wouldn't run for you.

**Comment by matejcik** [comment]
https://github.com/trezor/trezor-firmware/blob/master/tools/build_protobuf#L72-L84

**Comment by adonagy** [comment]
> The BITCOIN_ONLY check is added by a bunch of seds in tools/build_protobuf. make gen runs that. No idea why it wouldn't run for you.

Oh yeah, the seds. Always got a bunch of sed errors while running make gen, I should have checked into it. Looks like the problem was the sed on macOS. I did sort it out now. Everything is generating fine now. Thanks!

File: core/src/apps/tezos/sign_tx.py (hunk 8 above)
Comment (adonagy) [reply core/src/apps/tezos/sign_tx.py]: That was not intended. I'll fix it

**Comment by adonagy** [comment]
I've made the requested changes and fixed the generated files. If there is anything else, let me know.

**Comment by prusnak** [comment]
I squashed the changes into 384275ac732c8c08a824cef82ed4b6a76355b122 and added one more small refactor to 322417d9ed2a0bfb6efdc8784c23a55702a1d042. Both of the commits are in the master now.

@adonagy Can you please test that my changes did not break the functionality? The tests pass, but testing in the wild is best of course.

**Pull request closed without merging**